libr/arch/arch.c

```diff
@@ -0,0 +1,161 @@
+/* radare - LGPL - Copyright 2020 - pancake */
+
+#include <r_arch.h>
+#include <r_asm.h>
+#include "../config.h"
+
+static RArchPlugin *arch_static_plugins[] = { R_ARCH_STATIC_PLUGINS };
+
+static void plugin_free(RArchPlugin *p) {
+	if (p && p->fini) {
+		p->fini (NULL);
+	}
+}
+
+static bool is_valid(RArch *a, const char *name) {
+        RArchPlugin *h;
+        RListIter *iter;
+        if (!name || !*name) {
+                return false;
+        }
+        r_list_foreach (a->plugins, iter, h) {
+                if (!strcmp (h->name, name)) {
+                        return true;
+                }
+        }
+        return false;
+}
+
+R_API RArch *r_arch_new() {
+	RArch *a = R_NEW0 (RArch);
+	a->bits = R_SYS_BITS;
+	a->endian = R_SYS_ENDIAN_LITTLE;
+        a->plugins = r_list_newf ((RListFree)plugin_free);
+        if (!a->plugins) {
+                free (a);
+                return NULL;
+        }
+	size_t i;
+        for (i = 0; arch_static_plugins[i]; i++) {
+                r_arch_add (a, arch_static_plugins[i]);
+        }
+	return a;
+}
+
+R_API void r_arch_free(RArch *arch) {
+	free (arch);
+}
+
+// attributes
+
+static bool has_bits(RArchPlugin *h, RArchBits bits) {
+        return (h && h->bits && (bits & h->bits));
+}
+
+R_API bool r_arch_set_cpu(RArch *a, const char *cpu) {
+	// TODO: check if cpu is valid for the selected plugin
+	if (a) {
+		free (a->cpu);
+		a->cpu = cpu? strdup (cpu): NULL;
+	}
+	return true;
+}
+
+R_API bool r_arch_setup(RArch *a, const char *arch, RArchBits bits, RArchEndian endian) {
+	r_return_val_if_fail (a && arch, false);
+	bool ret = !r_arch_use (a, arch);
+	r_arch_set_endian (a, endian);
+	return ret | !r_arch_set_bits (a, bits);
+}
+
+R_API bool r_arch_set_syntax(RArch *a, int syntax) {
+	r_return_val_if_fail (a, false);
+	switch (syntax) {
+	case R_ASM_SYNTAX_REGNUM:
+	case R_ASM_SYNTAX_INTEL:
+	case R_ASM_SYNTAX_MASM:
+	case R_ASM_SYNTAX_ATT:
+	case R_ASM_SYNTAX_JZ:
+		a->syntax = syntax;
+		return true;
+	default:
+		return false;
+	}
+}
+
+R_API bool r_arch_set_endian(RArch *a, RArchEndian endian) {
+	r_return_val_if_fail (a, false);
+	switch (a->cur->endian) {
+	case R_SYS_ENDIAN_LITTLE:
+	case R_SYS_ENDIAN_BIG:
+	case R_SYS_ENDIAN_NONE:
+	case R_SYS_ENDIAN_BI:
+		a->endian = endian;
+		return true;
+	}
+	return false;
+}
+
+R_API bool r_arch_set_bits(RArch *a, int bits) {
+	r_return_val_if_fail (a, false);
+	if (has_bits (a->cur, bits)) {
+		a->bits = bits;
+		return true;
+	}
+	return false;
+}
+
+// plugins
+
+R_API bool r_arch_use(RArch *a, const char *name) {
+	r_return_val_if_fail (a && a->plugins, false);
+
+	if (!name || !*name) {
+		a->cur = NULL;
```

Comment:
yes, let me show you different situations:

* when building RArch without any plugin, it shouldnt crash, this happened already in 
* right now arch->cur == null means ther'es no arch plugin for the given arch because we are in a transitional state to let asm/anal use arch and their own plugins as fallback
* Once the transition is done ->cur null could mean that setup() or use() have failed. the thing is... if we do arch.use('404') what do u expect to happen? if you do RArch.new().cur what should it contain?

Reply:
I'm ok if this is as a transitional state, but after that I think cur should always be set.
When you do r_arch_new it can either automatically sets a default plugin, which can even be a dummy one (like all bytes are invalid, or something like that) or just sets NULL (but if you set NULL, that means that the RArch instance is not in the right state to be used yet). Then, if you do `r_arch_use("404")` and then someone calls `r_arch_encode()` it is doing it wrong! It means the user of the API did not check for errors before, which is bad, so having a `r_return_val_if_fail` on `a->cur` will detect those bad uses of the API, since you should always check for errors.